File: Ladder03/Java_code/BinaryTreeMaximumPathSum.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
public class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: An integer.
     */
    private int max = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        singleMaxPathSum(root);
        return max;
    }

    private int singleMaxPathSum(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int left = singleMaxPathSum(root.left);
        int right = singleMaxPathSum(root.right);
        int singleMaxPathSum = Math.max(root.val, Math.max(root.val + left, root.val + right));
        max = Math.max(max, Math.max(singleMaxPathSum, root.val + left + right));
        return singleMaxPathSum;
    }
}

File: Ladder03/Python_code/binary_tree_maximum_path_sum.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        this.val = val
        this.left, this.right = None, None
"""
class Solution:
    """
    @param root: The root of binary tree.
    @return: An integer
    """
    def maxPathSum(self, root):
        Max = [-sys.maxint]
        self.singleMax(root, Max)
        return Max[0]
    
    def singleMax(self, root, Max):
        if root is None:
            return 0
        left = self.singleMax(root.left, Max)
        right = self.singleMax(root.right, Max)
        singleMaxVal = 0
        singleMaxVal = max(root.val, root.val + left, root.val + right)
        Max[0] = max(Max[0], singleMaxVal, root.val + left + right)
        return singleMaxVal